### ISSUE TITLE
SVD and SVDpp fail to complete my matrix

### ISSUE THREAD

**Issue by lenhhoxung86** [NicolasHug/Surprise#205]
#### Description
Hello,

I'm using your library to complete my very sparse matrix and it appears to be working very well with basic methods such as KNN.
However, when I try using SVD and SVDpp, the results are very low compared to other methods. Moreover, changing parameters of SVD and SVDpp (learning rate, epochs, etc...) doesn't seem to change the output.
Could you please have a look and suggest me how can I improve the results for these methods?

Thank you very much.

#### Steps/Code to Reproduce
```
# load dataset
adj, features, train_mask, test_mask, mask_all = load_data() # these matrices are all sparse

features = features.toarray()
train_mask = train_mask.toarray()
test_mask = test_mask.toarray()
mask_all = mask_all.toarray()

# filter values
train_features = np.multiply(features,train_mask)
test_features = np.multiply(features,test_mask)

# format the data
train_dict = create_feature_dict(train_features)
testset = create_tuple_list(test_features)
# import pdb;pdb.set_trace()
df = pd.DataFrame(train_dict)

# A reader is still needed but only the rating_scale param is requiered.
value_range = find_value_range(train_features)

reader = Reader(rating_scale=value_range)

trainset = Dataset.load_from_df(df[['node_ids', 'hour_ids', 'values']], reader).build_full_trainset()


print('\nUSING NormalPredictor ===>')
# We'll use the famous SVD algorithm.
algo = NormalPredictor()
# Train the algorithm on the trainset, and predict ratings for the testset
algo.fit(trainset)
predictions = algo.test(testset)

# Then compute RMSE
accuracy.rmse(predictions)
accuracy.mae(predictions)

print('\nUSING BaselineOnly ===>')
# We'll use the famous SVD algorithm.
algo = BaselineOnly()
# Train the algorithm on the trainset, and predict ratings for the testset
algo.fit(trainset)
predictions = algo.test(testset)

# Then compute RMSE
accuracy.rmse(predictions)
accuracy.mae(predictions)

print('\nUSING KNNBasic ===>')
# We'll use the famous SVD algorithm.
algo = KNNBasic()
# Train the algorithm on the trainset, and predict ratings for the testset
algo.fit(trainset)
predictions = algo.test(testset)

# Then compute RMSE
accuracy.rmse(predictions)
accuracy.mae(predictions)

...
**print('\nUSING SVD ===>')
# We'll use the famous SVD algorithm.
algo = SVD() 
# Train the algorithm on the trainset, and predict ratings for the testset
algo.fit(trainset)
predictions = algo.test(testset)

# Then compute RMSE
accuracy.rmse(predictions)
accuracy.mae(predictions)

print('\nUSING SVDpp ===>')
# We'll use the famous SVD algorithm.
algo = SVDpp()
# Train the algorithm on the trainset, and predict ratings for the testset
algo.fit(trainset)
predictions = algo.test(testset)

# Then compute RMSE
accuracy.rmse(predictions)
accuracy.mae(predictions)**
```

#### Expected Results
Small values for RMSE and MAE for SVD and SVDpp

#### Actual Results
```
USING NormalPredictor ===>
RMSE: 60.7129
MAE:  49.7389

USING BaselineOnly ===>
Estimating biases using als...
RMSE: 31.0962
MAE:  24.4139

USING KNNBasic ===>
Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 34.9787
MAE:  24.8582

USING KNNBaseline ===>
Estimating biases using als...
Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 29.7501
MAE:  22.1741

USING KNNWithMeans ===>
Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 34.6566
MAE:  24.1420

USING KNNWithZScore ===>
Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 35.7003
MAE:  25.1922

USING SVD ===>
RMSE: 550.3130
MAE:  548.6182

USING SVDpp ===>
RMSE: 550.3130
MAE:  548.6182

USING NMF ===>
RMSE: 73.1111
MAE:  62.1129

USING SlopeOne ===>
RMSE: 33.8309
MAE:  23.7307

USING CoClustering ===>
RMSE: 35.4363
MAE:  26.8462
```

#### Versions
```
Linux-4.15.0-32-generic-x86_64-with-Ubuntu-16.04-xenial
Python 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609]
surprise 1.0.6
```

**Comment by NicolasHug** [comment]
It's hard giving feedback without any info about your dataset.

SVD and SVD++ are supposed to work very well on very sparse datasets such as the Netflix dataset.

Your rating scale seems to be very different from 1-5 which is where the default hyper parameters are the most effective. I think tuning the regularization and most importantly the learning rate give you much better results. You can use the GridSearch tools for that (you probably want to use a log scale for the learning rate)

**Comment by lenhhoxung86** [comment]
Yes, the rating scale is between 1.0 and 650.0 because the values are air pollution measurements. You suggested tuning the hyper parameters such as the learning rate and the regularization parameter. But I think it won't change the output. Concretely, I changed `lr_all` to `0.01` and `0.1` and the outputs were the same.
Furthermore, the SVD and SVDpp always gave me the same results.
What might be wrong here?

**Comment by NicolasHug** [comment]
They don't change because they probably don't converge. And I believe they don't converge because they have badly tuned learning rates (again, use grid search with a log scale. Trying just .01 and .1 is not enough)

SVD is a based on SGD (Stochastic Gradient Descent) which is sensitive to the scale of your data.

**Comment by NicolasHug** [comment]
Also, keep in mind that SVD models the "ratings" in a very specific way that is motivated by how users may interact with movies / objects.

I still don't know what your dataset is but if this model does not make sense for your data then it's normal that SVD gives bad results.

**Comment by lenhhoxung86** [comment]
just to confirm that using a learning rate of `1e-4` makes it work. Thanks for your help. Please close this issue.
